File: packages/pvm-assemblyscript/assembly/host-functions/general/export.ts
import { HostFunctionResult } from '../accumulate/base'
import { HostFunctionContext, HostFunctionParams, ExportParams } from './base'
import {
  ACCUMULATE_ERROR_CODES,
  GENERAL_FUNCTIONS,
  MAX_PACKAGE_EXPORTS,
  RESULT_CODES,
  SEGMENT_SIZE,
} from '../../config'
import { BaseHostFunction } from './base'

/**
 * EXPORT host function (Ω_E)
 *
 * Gray Paper Specification (pvm_invocations.tex, lines 529-544):
 * - Function ID: 7 (export)
 * - Gas Cost: 10
 * - Uses registers[7] for memory offset (p)
 * - Uses registers[8] for length, capped at Csegmentsize (4104)
 * - Creates zero-padded segment of exactly Csegmentsize
 * - Appends segment to export sequence
 * - Returns FULL if segoff + len(𝐞) >= Cmaxpackageexports (3072)
 * - Otherwise returns segoff + len(𝐞)
 */

export class ExportHostFunction extends BaseHostFunction {
  functionId: u64 = GENERAL_FUNCTIONS.EXPORT
  name: string = 'export'

  execute(
    context: HostFunctionContext,
    params: HostFunctionParams | null,
  ): HostFunctionResult {
    if (!params) {
      return new HostFunctionResult(RESULT_CODES.PANIC)
    }
    const exportParams = params as ExportParams
    if (!exportParams.refineContext) {
      return new HostFunctionResult(RESULT_CODES.PANIC)
    }

    // Gray Paper: p = registers[7]
    const memoryOffset = u32(context.registers[7])
    // Gray Paper: z = min(registers[8], Csegmentsize)
    const rawLength = u32(context.registers[8])
    const cappedLength =
      rawLength < SEGMENT_SIZE
        ? rawLength
        : SEGMENT_SIZE

    // Gray Paper: Check if Nrange[p][z] ⊆ readable[memory]
    const readableResult = context.ram.isReadableWithFault(
      memoryOffset,
      cappedLength,
    )
    const readable = readableResult.success
    const readableFaultAddress = readableResult.faultAddress
    if (!readable) {
      // Gray Paper: Return PANIC if memory not readable
      context.registers[7] = u64(0)
      return new HostFunctionResult(RESULT_CODES.PANIC)
    }

    // Gray Paper: Read data from memory
    const readResult = context.ram.readOctets(
      memoryOffset,
      cappedLength,
    )
    const data = readResult.data
    const readFaultAddress = readResult.faultAddress
    if (data === null || readFaultAddress !== 0) {
      context.registers[7] = u64(0)
      return new HostFunctionResult(RESULT_CODES.PANIC)
    }

    // Gray Paper: Create zero-padded segment of Csegmentsize
    const segment = this.createZeroPaddedSegment(data)

    // Gray Paper: Append to export sequence and check limits
    const result = this.appendToExports(exportParams, segment)

    if (result === -1) {
      // Gray Paper: Return FULL (2^64 - 5)
      context.registers[7] = ACCUMULATE_ERROR_CODES.FULL
    } else {
      // Gray Paper: Return segoff + len(𝐞)
      context.registers[7] = u64(result)
    }

    return new HostFunctionResult(255) // Continue execution
  }

  /**
   * Gray Paper: Create zero-padded segment of exactly Csegmentsize
   * zeropad{Csegmentsize}{mem[p:p+z]}
   */
  createZeroPaddedSegment(data: Uint8Array): Uint8Array {
    const segment = new Uint8Array(SEGMENT_SIZE)
    segment.set(data, 0)
    // Remaining bytes are already zero (Uint8Array initialization)
    return segment
  }

  /**
   * Gray Paper: Append segment to export sequence (𝐞)
   * Check if segoff + len(𝐞) >= Cmaxpackageexports
   * Return FULL if limit exceeded, otherwise return segoff + len(𝐞)
   */
  appendToExports(
    params: ExportParams,
    segment: Uint8Array,
  ): i64 {
    if (!params.refineContext) {
      return -1 // FULL indicator
    }
    const exportSegments = params.refineContext!.exportSegments
    const segoff = params.segmentOffset

    // Gray Paper: Check if segoff + len(𝐞) >= Cmaxpackageexports
    const currentLength = i64(exportSegments.length)
    if (segoff + currentLength >= i64(MAX_PACKAGE_EXPORTS)) {
      return -1 // FULL indicator
    }

    // Gray Paper: Append segment
    exportSegments.push(segment)

    // Gray Paper: Return segoff + len(𝐞)
    return segoff + currentLength
  }
}

File: packages/pvm-rust/src/host_functions/general/export.rs
//! EXPORT host function (Ω_E). Gray Paper: function ID 7.
//! r7=memory offset (p), r8=length (capped at SEGMENT_SIZE). No refineContext → PANIC.

use crate::config::{FUNC_EXPORT, REG_FULL, SEGMENT_SIZE};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

/// EXPORT (7): with refine_context: read memory, zero-pad segment, append; r7=segoff+len or FULL. No refine → PANIC.
pub struct ExportHostFunction;

impl HostFunction for ExportHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_EXPORT
    }
    fn name(&self) -> &'static str {
        "export"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let Some(refine) = context.refine_context.as_mut() else {
            return HostFunctionResult::panic();
        };

        let memory_offset = context.registers[7] as u32;
        let raw_length = context.registers[8] as u32;
        let capped_length = raw_length.min(SEGMENT_SIZE);

        let check = context.ram.is_readable_with_fault(memory_offset, capped_length);
        if !check.success {
            context.registers[7] = 0;
            return HostFunctionResult::panic();
        }
        let read_result = context.ram.read_octets(memory_offset, capped_length);
        let Some(data) = read_result.data else {
            context.registers[7] = 0;
            return HostFunctionResult::panic();
        };
        if read_result.fault_address != 0 {
            context.registers[7] = 0;
            return HostFunctionResult::panic();
        }

        let mut segment = vec![0u8; SEGMENT_SIZE as usize];
        let len = data.len().min(SEGMENT_SIZE as usize);
        segment[..len].copy_from_slice(&data[..len]);

        match refine.push_export_segment(segment) {
            Ok(result) => context.registers[7] = result as u64,
            Err(()) => context.registers[7] = REG_FULL,
        }
        HostFunctionResult::continue_execution()
    }
}
